services/api/service.go

```diff
@@ -2059,6 +2061,298 @@ func (api *RelayAPI) handleSubmitNewBlock(w http.ResponseWriter, req *http.Reque
 	w.WriteHeader(http.StatusOK)
 }
 
+func (api *RelayAPI) handleSubmitNewBlockV2(w http.ResponseWriter, req *http.Request) {
+	var pf common.Profile
+	var prevTime, nextTime time.Time
+
+	headSlot := api.headSlot.Load()
+	receivedAt := time.Now().UTC()
+	prevTime = receivedAt
+
+	args := req.URL.Query()
+	isCancellationEnabled := args.Get("cancellations") == "1"
+
+	log := api.log.WithFields(logrus.Fields{
+		"method":                "submitNewBlockV2",
+		"contentLength":         req.ContentLength,
+		"headSlot":              headSlot,
+		"cancellationEnabled":   isCancellationEnabled,
+		"timestampRequestStart": receivedAt.UnixMilli(),
+	})
+
+	// Log at start and end of request
+	log.Info("request initiated")
+	defer func() {
+		log.WithFields(logrus.Fields{
+			"timestampRequestFin": time.Now().UTC().UnixMilli(),
+			"requestDurationMs":   time.Since(receivedAt).Milliseconds(),
+		}).Info("request finished")
+	}()
+
+	// If cancellations are disabled but builder requested it, return error
+	if isCancellationEnabled && !api.ffEnableCancellations {
+		log.Info("builder submitted with cancellations enabled, but feature flag is disabled")
+		api.RespondError(w, http.StatusBadRequest, "cancellations are disabled")
+		return
+	}
+
+	var err error
+	var r io.Reader = req.Body
+	if req.Header.Get("Content-Encoding") == "gzip" {
+		r, err = gzip.NewReader(req.Body)
+		if err != nil {
+			log.WithError(err).Warn("could not create gzip reader")
+			api.RespondError(w, http.StatusBadRequest, err.Error())
+			return
+		}
+		log = log.WithField("gzip-req", true)
+	}
+
+	var buf bytes.Buffer
+	rHeader := io.TeeReader(r, &buf)
+
+	// Header at most 944 bytes.
+	headBuf := make([]byte, 944)
+
+	// Read header bytes.
+	_, err = io.ReadFull(rHeader, headBuf)
+	if err != nil {
+		log.WithError(err).Warn("could not read full header")
+		api.RespondError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	// Unmarshall just header.
+	var header common.SubmitBlockRequestV2Optimistic
+	err = header.UnmarshalSSZHeaderOnly(headBuf)
+	if err != nil {
+		log.WithError(err).Warn("could not unmarshall request")
+		api.RespondError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	nextTime = time.Now().UTC()
+	pf.Decode = uint64(nextTime.Sub(prevTime).Microseconds())
+	prevTime = nextTime
+
+	bid := header.Message
+	sig := header.Signature
+	eph := header.ExecutionPayloadHeader
+
+	log = log.WithFields(logrus.Fields{
+		"timestampAfterDecoding": time.Now().UTC().UnixMilli(),
+		"slot":                   bid.Slot,
+		"builderPubkey":          bid.BuilderPubkey.String(),
+		"blockHash":              bid.BlockHash.String(),
+		"proposerPubkey":         bid.ProposerPubkey.String(),
+		"parentHash":             bid.ParentHash.String(),
+		"value":                  bid.Value.String(),
+	})
+
+	// Verify the signature
```

Comment:
Is there a reason why the signature verification was moved above some of the other checks? (compared to handleSubmitNewBlock v1)

Reply:
honestly it was just a gut feeling that the signature check should probably come first.

Reply:
good rule of thumb to do inexpensive checks first, if they don't lead to DoS issues and can quickly determine if the request is bad or not

then ramp up in resource demand to signature verification

Reply:
yep agree, signature checks are expensive and should come after the cheaper checks

Reply:
done in https://github.com/flashbots/mev-boost-relay/pull/524/commits/3e5f962b69ac7753308640d9de89b413fc53d407 !

ty